cpp/ql/lib/semmle/code/cpp/models/implementations/GetText.qll

```diff
@@ -0,0 +1,35 @@
+import semmle.code.cpp.models.interfaces.DataFlow
+
+/**
+ * Returns the transated text index for a given gettext function `f`
+ */
+private int getTextArg(Function f) {
+  // basic variations of gettext
+  f.hasGlobalOrStdName("gettext") and result = 0
+  or
+  f.hasGlobalOrStdName("dgettext") and result = 1
+  or
+  f.hasGlobalOrStdName("dcgettext") and result = 1
+  or
+  // plural variations of gettext that take one format string for singular and another for plural form
+  f.hasGlobalOrStdName("ngettext") and
+  (result = 0 or result = 1)
+  or
+  f.hasGlobalOrStdName("dngettext") and
+  (result = 1 or result = 2)
+  or
+  f.hasGlobalOrStdName("dcngettext") and
+  (result = 1 or result = 2)
+}
+
+class GetTextFunction extends DataFlowFunction {
```

Comment:
I think this should extend `TaintFlowFunction` rather than `DataFlowFunction`, and override `hasTaintFlow` rather than `hasDataFlow`.  Data flow is intended for cases where the source value is directly used as or copied to the destination, whereas taint is for when the source value substantially influences the destination (in a way that could create a vulnerability).

Reply:
This was a debate I had with @MathiasVP . This is one I really would argue should absolutely be DataFlow, not taint. I think @MathiasVP at least was sympathetic to my argument, so we are hoping a DCA test shows the thesis is correct that it will minimally impact anyone, whereas failing to model it this way would result in lots of false negatives generally.

Reply:
Ah, so if I'm understanding you it's for pragmatic reasons - we get better results with this as a dataflow model.

Reply:
> whereas failing to model it this way would result in lots of false negatives generally.

While I disagree with the severity of this conclusion I do agree that this function is a great example of the difficulty in semantically distinguishing between when a model should be _data_ flow and when it needs to be _taint_ flow. On many systems a call to `gettext` will be equivalent to a call to `strdup` (which is a dataflow model), but on other systems the message will be completely changed because is actually does a translation.

Reply:
I'm still waiting for @bdrodes to chime in on this, because I'm not all that familiar with `gettext`, but I'm increasingly thinking of it as doing something like:
```
char *gettext(char *key) {
  return strdup(some_condition ? language_database_lookup(key) : key);
}
```
If this is accurate, we absolutely do have dataflow, just as we would for `x` in `some_condition ? 0 : x`.

Reply:
Yeah, that's my mental model of the function as well

Reply:
That is my mental model as well. I would need to pull the gnu source to see what is actually done, but effectively that's exactly what happens (at least in my understanding of the library)